service/worker/scanner/shardscanner/activities.go

```diff
@@ -0,0 +1,336 @@
+	ShardCorruptKeysQuery = "shard_corrupt_keys"
+)
+
+// ScannerConfigActivity will read dynamic config, apply overwrites and return a resolved config.
+func ScannerConfigActivity(
+	activityCtx context.Context,
+	params ScannerConfigActivityParams,
+) (ResolvedScannerWorkflowConfig, error) {
+	ctx := activityCtx.Value(params.ContextKey).(Context)
+	dc := ctx.Config.DynamicParams
+
+	result := ResolvedScannerWorkflowConfig{
+		GenericScannerConfig: GenericScannerConfig{
+			Enabled:                 dc.Enabled(),
+			Concurrency:             dc.Concurrency(),
+			PageSize:                dc.PageSize(),
+			BlobstoreFlushThreshold: dc.BlobstoreFlushThreshold(),
+			ActivityBatchSize:       dc.ActivityBatchSize(),
+		},
+	}
+
+	if ctx.Hooks != nil && ctx.Hooks.Config != nil {
+		result.CustomScannerConfig = ctx.Hooks.Config(ctx)
+	}
+
+	overwrites := params.Overwrites.GenericScannerConfig
+	if overwrites.Enabled != nil {
+		result.GenericScannerConfig.Enabled = *overwrites.Enabled
+	}
+	if overwrites.Concurrency != nil {
+		result.GenericScannerConfig.Concurrency = *overwrites.Concurrency
+	}
+	if overwrites.PageSize != nil {
+		result.GenericScannerConfig.PageSize = *overwrites.PageSize
+	}
+	if overwrites.BlobstoreFlushThreshold != nil {
+		result.GenericScannerConfig.BlobstoreFlushThreshold = *overwrites.BlobstoreFlushThreshold
+	}
+
+	if overwrites.ActivityBatchSize != nil {
+		result.GenericScannerConfig.ActivityBatchSize = *overwrites.ActivityBatchSize
+	}
+
+	if params.Overwrites.CustomScannerConfig != nil {
+		result.CustomScannerConfig = *params.Overwrites.CustomScannerConfig
+	}
+
+	return result, nil
+}
+
+// ScanShardActivity will scan a collection of shards for invariant violations.
+func ScanShardActivity(
+	activityCtx context.Context,
+	params ScanShardActivityParams,
+) ([]ScanReport, error) {
+	heartbeatDetails := ScanShardHeartbeatDetails{
+		LastShardIndexHandled: -1,
+		Reports:               nil,
+	}
+	if activity.HasHeartbeatDetails(activityCtx) {
+		if err := activity.GetHeartbeatDetails(activityCtx, &heartbeatDetails); err != nil {
+			return nil, err
+		}
+	}
+	for i := heartbeatDetails.LastShardIndexHandled + 1; i < len(params.Shards); i++ {
+		currentShardID := params.Shards[i]
+		shardReport, err := scanShard(activityCtx, params, currentShardID, heartbeatDetails)
+		if err != nil {
+			return nil, err
+		}
+		heartbeatDetails = ScanShardHeartbeatDetails{
+			LastShardIndexHandled: i,
+			Reports:               append(heartbeatDetails.Reports, *shardReport),
+		}
+	}
+	return heartbeatDetails.Reports, nil
+}
+
+func scanShard(
+	activityCtx context.Context,
+	params ScanShardActivityParams,
+	shardID int,
+	heartbeatDetails ScanShardHeartbeatDetails,
+) (*ScanReport, error) {
+	ctx := activityCtx.Value(params.ContextKey).(Context)
+	scope := ctx.Scope.Tagged(metrics.ActivityTypeTag(params.ContextKey.String()))
```

Comment:
Does params.ContextKey.String() identify the scan type or this activity name?

Reply:
It seems like we want a tag called scan - or maybe a subscope created at the root each scanner to the separation between different scanner's metrics is clear.

Reply:
Yes, this is an actual bug. I need to change ActivityType to actual activity name, will add workflow type to distinguish between scanners. Will add check to tests as well.

Regarding subscope: our `metrics` scope does not implement SubScope interface method.